packages/dds/tree/src/feature-libraries/indexing/anchorTreeIndex.ts

```diff
@@ -0,0 +1,285 @@
+ * A key can map to multiple nodes but each collection of nodes only results in a single value.
+ *
+ * @remarks
+ * Detached nodes are stored in the index but filtered out when any public facing apis are called. This means that
+ * calling {@link keys} will not include any keys that are stored in the index but only map to detached nodes.
+ *
+ * TODO: need to make sure key finders are deterministic or have a way to invalidate them
+ */
+export class AnchorTreeIndex<TKey extends TreeValue, TValue>
+	implements TreeIndex<TKey, TValue>
+{
+	/**
+	 * Caches {@link KeyFinder}s for each schema definition. If a schema maps to null, it does not
+	 * need to be considered at all for this index. This allows us to skip subtrees that aren't relevant
+	 * as a performance optimization.
+	 */
+	private readonly keyFinders = new Map<TreeNodeSchemaIdentifier, KeyFinder<TKey> | null>();
+	/**
+	 * The actual index from keys to anchor nodes.
+	 */
+	private readonly nodes = new Map<TKey, TreeIndexNodes<AnchorNode>>();
+	/**
+	 * Keeps track of anchors for disposal.
+	 */
+	private readonly anchors = new Map<AnchorNode, Anchor>();
+
+	/**
+	 * @param forest - the forest that is being indexed
+	 * @param indexer - a function that retrieves the key finder based on a given schema or undefined if the schema does not have an associated key finder
+	 * @param getValue - a function that returns the value or undefined given at least one anchor node
+	 * @param checkTreeStatus - a function that gets the tree status from an anchor node, used for filtering out detached nodes
+	 */
+	public constructor(
+		private readonly forest: IForestSubscription,
+		indexer: (schemaId: TreeNodeSchemaIdentifier) => KeyFinder<TKey> | undefined,
+		private readonly getValue: (anchorNodes: TreeIndexNodes<AnchorNode>) => TValue | undefined,
+		private readonly checkTreeStatus: (node: AnchorNode) => TreeStatus | undefined,
+	) {
+		/**
+		 * Given a cursor in field mode, recursively indexes all nodes under the field.
+		 */
+		const indexField = (fieldCursor: ITreeSubscriptionCursor): void => {
+			forEachNode(fieldCursor, (nodeCursor) => {
+				const keyFinder = getOrCreate(
+					this.keyFinders,
+					fieldCursor.type,
+					// if the indexer does not return a key finder for this schema, we cache a null value to indicate the indexer
+					// does not need to be called if this schema is encountered in the future
+					(schema) => indexer(schema) ?? null,
+				);
+
+				if (keyFinder !== null) {
+					const key = keyFinder(fieldCursor);
+					const anchor = fieldCursor.buildAnchor();
+					const anchorNode = forest.anchors.locate(anchor) ?? fail("expected anchor node");
+
+					const nodes = this.nodes.get(key);
+					if (nodes !== undefined) {
+						// if the key already exists in the index, the anchor node is appended to its list of nodes
+						this.nodes.set(key, [...nodes, anchorNode]);
```

Comment:
any reason to clone the array rather than push to it?

Reply:
the array is readonly for api purposes and gets super ugly (not even sure if it's completely feasible) to make a bunch of casts

Reply:
don't we just need the output type (the thing we hand to getValue) to implement readonly array...which array does? can we not just type our internal type as arry, e.g. `private readonly nodes = new Map<TKey, AnchorNode[]>();`

Reply:
we filter the array (meaning allocate a new one) before handing it to getValue anyway, right?

Reply:
oh yeah, you're right, I was just doing something stupid when I tried it the first time